### ISSUE TITLE
Add support for deploying to GitHub pages with the github personal access token

### ISSUE THREAD

**Issue by d0ugal** [mkdocs/mkdocs#632]
This would make deploying after a CI build really easy.

Something like http://ablog.readthedocs.org/manual/auto-github-pages-deploys/

**Comment by waylan** [comment]
That's cool. Although perhaps not as useful to a documentation engine as it would be to a blogging engine. Unless, of course, the documentation included multiple versions and only the "dev" version of the docs was updated on every push. I would expect that "release" versions of docs would generally only be updated upon release. That said, various users could have different non-standard use cases that this might help with. I just don't see it as a high priority without multiple version support. Still, very cool and worth linking to so as not to forget about it for later.

**Comment by waylan** [comment]
One possible concern. I'm assuming the token is set as an environment variable on the user's CI account (which is only ever referenced via the variable within scripts/config files) as a security measure. That way the token is never in a public repo or in the publicly readable CI build logs (only the variable name). Of course, that is a significant part of what makes this cool. But, what about Pull Requests? If a PR is made (even if the user making the pull request has no CI account), a build is run by CI on the PR using the parent repo's account. Will these builds have access to the environment variable? I would expect not (otherwise any PR could overwrite your website--potentially in malicious ways), but I would want to confirm that before implementing.

And if they don't, how do the tests show that they pass on the CI build as the missing environment variable would cause an error? Every PR showing as failed for this reason alone would get annoying real fast. Maybe this has all been accounted for and is a non-problem, but I would want to check into it first.

**Comment by d0ugal** [comment]
> the documentation included multiple versions and only the "dev" version of the docs was updated on every push.

Yup, that is exactly what I'd imagined. So this would be dependent on #193 for MkDocs itself. It is something that I would want once we have multiple version builds.

re: your concerns about the workflow, yeah, that needs looking into. Travis CI has options to handle these problems but I'm not totally familiar with them.

**Comment by gsouf** [comment]
:+1:

**Comment by d0ugal** [comment]
I was able to do this fairly easily with `mkdocs-bootswatch` in the [Travis config](https://github.com/mkdocs/mkdocs-bootswatch/blob/master/.travis.yml#L29-L35). Most of the setup is just git commands. So I don't think there is as much for us to simplify as I originally thought.

    mkdocs gh-deploy --github-remote "https://${GH_TOKEN}@github.com/mkdocs/mkdocs-bootswatch.git"

With this command, we could create a temporary remote with the URI provided. Fetch the gh-pages branch and then deploy to that remote. Users would need to globally setup their git author and email for that to work well. We could just document this.

Alternatively we could try and do some more magic.

    mkdocs gh-deploy --github-token "${GH_TOKEN}"

With this approach we can figure out the rest of the remote and combine it with the token. This is slightly nicer for users, but I am not sure it is worth the extra effort.

Either way, this has become far lower priority for me now I've found a good way to work with what we have. It might even be worth properly documenting what is currently possible for now.

**Comment by gsouf** [comment]
To me it looks a top feature

**Comment by xen0l** [comment]
I'd like this, too.

**Comment by bameda** [comment]
I like the approach in Lektor to do this  https://www.getlektor.com/docs/deployment/travisci/

**Comment by michaeltlombardi** [comment]
Just as another point in favor of this type of utility, we include documentation with every project and publish to internal static sites _every_ time the docs are updated via CI.

We don't have 'released' version so much as 'always up to date' version for most of our docs (the exception being downloadable artifacts rather than services/processes etc), so continually pushing the updates is 💯 for us.

**Comment by d0ugal** [comment]
It is pretty easy to do now, but we should make it easier.

https://github.com/d0ugal/retrace/blob/master/.travis.yml#L32-L37

Lektor does look like it does it well.

**Comment by bameda** [comment]
yes, I use this for my script https://github.com/bameda/python-madrid-learn-doc/blob/master/.travis.yml#L13-L18 too

**Comment by 9cibot** [comment]
+ 1

**Comment by oprypin** [comment]
With availability (and popularity) of GitHub Actions, it is now much better to just use `mkdocs build` and then trust the deployment of the `site` dir to one of the GitHub-specific "actions", all of which should also expose the ability to use the token.

* [JamesIves/github-pages-deploy-action](https://github.com/JamesIves/github-pages-deploy-action)
* [crazy-max/ghaction-github-pages](https://github.com/crazy-max/ghaction-github-pages)
* [peaceiris/actions-gh-pages](https://github.com/peaceiris/actions-gh-pages)
* [oprypin/push-to-gh-pages](https://github.com/oprypin/push-to-gh-pages)

And for local usage, a SSH key or interactive auth work. Or indeed incorporating the token into the remote URL as in https://github.com/mkdocs/mkdocs/issues/632#issuecomment-261793168

So in my view, this doesn't need to be implemented.

**Comment by oprypin** [comment]
If someone disagrees with the closure, further comments are welcome.
